lang/funcs/core/coresys/cpucount_fact.go

```diff
@@ -0,0 +1,218 @@
+	socketFile      = "pipe.sock"
+	cpuDevpathRegex = "/devices/system/cpu/cpu[0-9]"
+)
+
+func init() {
+	facts.ModuleRegister(moduleName, "cpu_count", func() facts.Fact { return &CPUCountFact{} }) // must register the fact and name
+}
+
+// CPUCountFact is a fact that returns the current CPU count.
+type CPUCountFact struct {
+	init      *facts.Init
+	closeChan chan struct{}
+}
+
+// Info returns static typing info about what the fact returns.
+func (obj *CPUCountFact) Info() *facts.Info {
+	return &facts.Info{
+		Output: types.NewType("int"),
+	}
+}
+
+// Init run startup code for this fact. Initializes the closeChan and sets the
+// facts.Init variable.
+func (obj *CPUCountFact) Init(init *facts.Init) error {
+	obj.init = init
+	obj.closeChan = make(chan struct{})
+	return nil
+}
+
+// Stream returns the changing values that this fact has over time. It will
+// first poll sysfs to get the initial cpu count, and then receieves UEvents
+// from the kernel as CPUs are added/removed.
+func (obj CPUCountFact) Stream() error {
+	defer close(obj.init.Output) // signal when we're done
+
+	ss, err := socketset.NewSocketSet(rtmGrps, socketFile, unix.NETLINK_KOBJECT_UEVENT)
+	if err != nil {
+		return errwrap.Wrapf(err, "error creating socket set")
+	}
+
+	// waitgroup for netlink receive goroutine
+	wg := &sync.WaitGroup{}
+	defer wg.Wait()
+	defer ss.Close()
+	defer ss.Shutdown()
+
+	eventChan := make(chan *nlChanEvent) // updated in goroutine when we receive uevent
+	closeChan := make(chan struct{})     // channel to unblock selects in goroutine
+	defer close(closeChan)
+
+	// Start waiting for kernel to poke us about new device changes on the system
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		defer close(eventChan)
+		for {
+			uevent, err := ss.ReceiveUEvent() // calling Shutdown will stop this from blocking
+			if obj.init.Debug {
+				obj.init.Logf("sending uevent SEQNUM: %s", uevent.Data["SEQNUM"])
+			}
+			select {
+			case eventChan <- &nlChanEvent{
+				uevent: uevent,
+				err:    err,
+			}:
+			case <-closeChan:
+				return
+			}
+		}
+	}()
+
+	startChan := make(chan struct{})
+	close(startChan) // trigger the first event
+	var cpuCount, newCount int64 = 0, -1
+	for {
+		select {
+		case <-startChan:
+			startChan = nil // disable
+			newCount, err = getCPUCount()
+			if err != nil {
+				obj.init.Logf("could not get initial CPU count. Setting to zero.")
+			}
+		case event, ok := <-eventChan:
+			if !ok {
+				continue
+			}
+			if event.err != nil {
+				return errwrap.Wrapf(event.err, "error receiving uevent")
+			}
+			if obj.init.Debug {
+				obj.init.Logf("received uevent SEQNUM: %s", event.uevent.Data["SEQNUM"])
+			}
+			if isCPUEvent(event.uevent) {
+				newCount, err = getCPUCount()
+				if err != nil {
+					obj.init.Logf("could not getCPUCount: %e", err)
+					continue
+				}
+			}
+		case <-obj.closeChan:
+			return nil
+		}
+
+		if newCount == cpuCount {
+			continue
+		}
+		cpuCount = newCount
+		st := &types.IntValue{
```

Comment:
no need to use a var, just put it in directly... and this var name was copied from when it was a struct ;)